node/subsystem-util/src/rolling_session_window.rs

```diff
@@ -94,55 +100,172 @@ pub enum SessionWindowUpdate {
-			Err(kind) => Err(SessionsUnavailable {
-				kind,
-				info: Some(SessionsUnavailableInfo {
-					window_start,
-					window_end: session_index,
-					block_hash,
+		// Fetch session information from DB.
+		let maybe_stored_window = Self::db_load(db_params.clone());
+
+		// Get the DB stored sessions and recompute window start based on DB data.
+		let (mut window_start, stored_sessions) =
+			if let Some(mut stored_window) = maybe_stored_window {
+				// Check if DB is ancient.
+				if earliest_non_finalized_block_session >
+					stored_window.earliest_session + stored_window.session_info.len() as u32
+				{
+					// If ancient, we scrap it and fetch from chain state.
+					stored_window.session_info.clear();
+				}
+
+				// The session window might extend beyond the last finalized block, but that's fine as we'll prune it at
+				// next update.
+				let window_start = if stored_window.session_info.len() > 0 {
+					// If there is at least one entry in db, we always take the DB as source of truth.
+					stored_window.earliest_session
+				} else {
+					on_chain_window_start
+				};
+
+				(window_start, stored_window.session_info)
+			} else {
+				(on_chain_window_start, Vec::new())
+			};
+
+		// Try to load sessions from DB first and load more from chain state if needed.
+		let sessions_missing_count = session_index
+			.saturating_sub(window_start)
+			.saturating_add(1)
+			.saturating_sub(stored_sessions.len() as u32);
+
+		// Extend from chain state.
+		let sessions = if sessions_missing_count > 0 {
+			match extend_sessions_from_chain_state(
+				stored_sessions,
+				&mut sender,
+				block_hash,
+				&mut window_start,
+				session_index,
+			)
+			.await
+			{
+				Err(kind) => Err(SessionsUnavailable {
+					kind,
+					info: Some(SessionsUnavailableInfo {
+						window_start,
+						window_end: session_index,
+						block_hash,
+					}),
 				}),
-			}),
-			Ok(s) => Ok(Self { earliest_session: window_start, session_info: s, window_size }),
+				Ok(sessions) => Ok(sessions),
+			}?
+		} else {
+			// There are no new sessions to be fetched from chain state.
+			Vec::new()
+		};
+
+		Ok(Self {
+			earliest_session: window_start,
+			session_info: sessions,
+			window_size,
+			db_params: Some(db_params),
+		})
+	}
+
+	// Load session information from the parachains db.
+	fn db_load(db_params: DatabaseParams) -> Option<StoredWindow> {
+		match db_params.db.get(db_params.db_column, STORED_ROLLING_SESSION_WINDOW).ok()? {
+			None => None,
+			Some(raw) => {
+				let maybe_decoded = StoredWindow::decode(&mut &raw[..]).map(Some);
+				match maybe_decoded {
+					Ok(decoded) => decoded,
+					Err(err) => {
+						gum::warn!(
+							target: LOG_TARGET,
+							?err,
+							"Failed decoding db entry; will start with onchain session infos and self-heal DB entry on next update."
+						);
+						None
+					},
+				}
+			},
+		}
+	}
+
+	// Saves/Updates all sessions in the database.
+	fn db_save(&mut self, stored_window: StoredWindow) {
```

Comment:
> So we have as of now 2 subsystems which keep track of the same thing and then both write to the DB the same or a similar thing. While the caching per subsystem makes sense, having each of them write the same things to DB, does not seem great.

I reasoned about this and my logic was that the DB writing happens only on session change, which is a rare event, for which I don't think there would be any discrepancy across the 2 subsystems via active leaves update signal. The only thing that I was concerned about is the window size, which might get different over time (currently is 6 for both). 
Currently, the DB is read only when constructing the rolling session window (at node startup), while anything that comes later will work with the data in memory.

> 
> Ideally this would be managed by something like a subsytem (could be a part of runtime-api subystem or its own), which then is the source of truth. This subsystem would keep itself up2date with regards to sessions and makes sure to persist sessions to db. The `RollingSessionWindow` would then be a local cache that requests data from this subsystem ... Given that the window logic would then reside in that subsystem, the local cache could actually equally well be handled by [`RuntimeInfo`](https://github.com/paritytech/polkadot/blob/efb82ef7b0a0f0a67b43b616471db5921f45a614/node/subsystem-util/src/runtime/mod.rs#L61).
> 

The downside I can  see with this subsystem approach is more code/complexity, but indeed it would provide a single source of truth as in a subsystem. What I have in mind to keep things simple is an `allow_write` parameter for the RollingSessionWindow which would be set to false for `approval-voting` for example,  while `dispute-coordinator` would set it to `true`. However, this flag approach I am talking about would be less obvious to anyone unfamiliar with the code. 

> We could then also re-route all sessioninfo requests to that subsystem, so all subsystems are guaranteed to fetch the same stuff - this would also ensure the invariant we should really have, that if something can be backed, it can also be disputed.

To recap: Only reading from db at startup, removing the window size parameter and the guarantees of eventual consistency of overseer signal processing in both subsystems should provide the same invariant. In extreme circumstances, like for example one subsystem getting stuck we would have probably an older window saved to DB, but that's something which is immediately recovered on node startup, assuming the node doesn't go offiline for too long (which is something that will force fetching sessions from chain state in any approach)

I'm a bit on the fence about  this, reworking this to a different subsystem as you described would mean quite a lot of additional work wich in practice would produce the same results. The only thing that makes me think to implement it as a subsystem is to have a crisper design and more obvious to someone new looking to the code.

WDYT ?

Reply:
Hmm. I think I should just make the current rolling session read only. And move the whole db update/writing to an existing subsystem, maybe runtime api is a good home for this. Based on what I said above, we shouldnt be needing to query that subsystem, since we would keep everything cached and only load db at startup.

Reply:
Hmm, having a subsystem only concerned with writes seems weird. I agree with your argument that it should be fine for now as is. Longer term, the dedicated subsystem with local caches in other subsystems seems to be the most clean solution to me. => We would have a guarantee that all subsystem are seeing the same thing/are resilient to the same errors which would be a good overall improvement, not only for approval voting and disputes.

Reply:
Yeah, not really a good idea to have one subsystem just for syncing the session window to disk, but cant find a really good place for this purpose.

I'll write an issue for the longer term changes as discussed above. I think also most of the effort will go into fixing tests in approval voting and dispute coordinator rather than the actual implementation.

Reply:
Please reference the ticket in a TODO in the code as well. Helps keeping track.

Reply:
https://github.com/paritytech/polkadot/issues/6144